src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -108,18 +108,33 @@ private static unsafe SafeFileHandle CreateFile(string fullPath, FileMode mode,
 
         private static unsafe void Preallocate(string fullPath, long preallocationSize, SafeFileHandle fileHandle)
         {
-            // preallocationSize must be ignored for non-seekable files, unsupported file systems
-            // and other failures other than ERROR_DISK_FULL and ERROR_FILE_TOO_LARGE
-            if (!FileStreamHelpers.TrySetFileLength(fileHandle, preallocationSize, out int errorCode)
-                && errorCode == Interop.Errors.ERROR_DISK_FULL || errorCode == Interop.Errors.ERROR_FILE_TOO_LARGE)
-            {
-                // Windows does not modify the file size if the request can't be satisfied in atomic way.
-                // posix_fallocate (Unix) implementation might consume all available disk space and fail after that.
-                // To ensure that the behaviour is the same for every OS (no incomplete or empty file), we close the handle and delete the file.
+            var allocationInfo = new Interop.Kernel32.FILE_ALLOCATION_INFO
+            {
+                AllocationSize = preallocationSize
+            };
+
+            if (!Interop.Kernel32.SetFileInformationByHandle(
+                fileHandle,
+                Interop.Kernel32.FileAllocationInfo,
+                &allocationInfo,
+                (uint)sizeof(Interop.Kernel32.FILE_ALLOCATION_INFO)))
+            {
+                int errorCode = Marshal.GetLastPInvokeError();
+
+                // we try to mimic the atomic NtCreateFile here:
+                // if preallocation fails, close the handle and delete the file
                 fileHandle.Dispose();
                 Interop.Kernel32.DeleteFile(fullPath);
 
-                throw new IOException(SR.Format(errorCode == Interop.Errors.ERROR_DISK_FULL ? SR.IO_DiskFull_Path_AllocationSize : SR.IO_FileTooLarge_Path_AllocationSize, fullPath, preallocationSize));
+                switch (errorCode)
+                {
+                    case Interop.Errors.ERROR_DISK_FULL:
+                        throw new IOException(SR.Format(SR.IO_DiskFull_Path_AllocationSize, fullPath, preallocationSize));
+                    case Interop.Errors.ERROR_FILE_TOO_LARGE:
+                        throw new IOException(SR.Format(SR.IO_FileTooLarge_Path_AllocationSize, fullPath, preallocationSize));
+                    default:
+                        throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
```

Comment:
If we consider that Unix should not throw with errors other than ENOSPC and EFBIG, we should consider the same for Windows, right?

Reply:
Right.